### PR TITLE
add PSACT custom resource

### PR DESCRIPTION
**Pull request by jiaqiluo**
## Issue: 

 
https://github.com/rancher/terraform-provider-rancher2/issues/1189
## Problem

 
Rancher v2.7.2 supports a new CRD for [Pod Security Admission (PSA) Configuration Templates](https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/psa-config-templates) that can only be created outside of TF manually.

## Solution

 
Add a new resource and data source named  `rancher2_pod_security_admission_configuration_template`  for [rancher2_pod_security_admission_configuration_template](https://github.com/rancher/rancher/blob/release/v2.7/pkg/client/generated/management/v3/zz_generated_pod_security_admission_configuration_template.go?rgh-link-date=2023-08-03T18%3A34%3A57Z) so users can manage custom admission configuration templates via Terraform for RKE and RKE2/K3s clusters.

### Manual Testing

- Tested creating, modifying, and removing a custom PSACT resource. All tests are successful. 
- Tested creating a template with bad data and this failed as expected.
- Tested creating and destroying a PSACT along with an RKE2 cluster and RKE1 cluster. All tests are successful. 
- Tested fetching an existing `rancher2_pod_security_admission_configuration_template` data source

### Automated Testing

**unit test**

The standard sets of tests are added for the new `Resource` & `DataSource`. 

**acceptance test**

The acceptance tests are added but "disabled" by making the function names lowercase. This is because the current framework does not support running acceptance tests in Rancher v2.7.9 and v2.8.1. 

Here is how the acceptance test works:
- step 1: start a k3s cluster, currently using the version`v1.19.10-k3s1` 
- step 2: deploy Rancher v2.3.6 using Helm 
- step 3: create a set of Resources via TF into the Rancher setup, and validate the expected values present
- step 4: upgrade  Rancher to 2.5.9 using Helm

Steps 3 and 4 are repeated to upgrade and test on Rancher v2.4.13,  v2.5.9, and v2.6.2. 

To add the tests for the PSACT resource, I tried to add Rancher v2.7.9 and v2.8.1 to the version list. However, it did not work due to the following reasons:
 - the local cluster version, v1.19.10-k3s1, is too old for Rancher v2.7.9 and v2.8.1, as well as a newer version of Cert-manager, to run
 - Some legacy resources, such as rancher2_app, can be not deployed in Rancher v2.7.2 and above. More specifically the app used in the test is deprecated and no new chart is available.  

To make the acceptance test work, we need to:
- add support for upgrading the local cluster
- add support for deploying new versions of Rancher and Cert-manger
- filter out the broken tests on the newer version of Rancher 

That work is large enough to be a separate task. (Update: the issue was made https://github.com/rancher/terraform-provider-rancher2/issues/1308) 

The good news is that the manual and automated tests that QA performs can cover more than the acceptance I planned to add. 

## QA Testing Considerations

 
- create, update, and delete `rancher2_pod_security_admission_configuration_template` 
- delete a `rancher2_pod_security_admission_configuration_template` resource when it is used by existing clusters
- create and delete a `rancher2_pod_security_admission_configuration_template`  and an RKE2 or RKE1 cluster that uses it in the same TF configuration
- fetch an existing `rancher2_pod_security_admission_configuration_template` data source

### Regressions Considerations

The new resource and data source themselves should not affect anything existing. 

QA should validate there is no regression for deleting a `rancher2_cluster_v2` resource because one change is made in this PR.

### CODE REVIEW

File: rancher2/resource_rancher2_cluster_v2.go [hunk 1]
```diff
@@ -202,6 +202,28 @@ func resourceRancher2ClusterV2Delete(d *schema.ResourceData, meta interface{}) e
...
+			Pending:    []string{"removing"},
+			Target:     []string{"removed"},
+			Refresh:    clusterStateRefreshFunc(client, v1ClusterName),
+			Timeout:    d.Timeout(schema.TimeoutDelete),
+			Delay:      1 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+		_, waitErr = stateConf.WaitForState()
+		if waitErr != nil {
+			return fmt.Errorf("[ERROR] waiting for cluster (%s) to be removed: %s", cluster.ID, waitErr)
```
Comment (bfbachmann) [inline rancher2/resource_rancher2_cluster_v2.go]: Minor: could wrap this error with %w. Same goes for the other errors created with `fmt.Errorf` below in this PR.
```suggestion
			return fmt.Errorf("[ERROR] waiting for cluster (%s) to be removed: %w", cluster.ID, waitErr)
```

File: rancher2/resource_rancher2_pod_security_admission_configuration_template.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	log.Printf("[INFO] Deleting PodSecurityAdmissionConfigurationTemplate with ID %s", id)
+	client, err := meta.(*Config).ManagementClient()
+	if err != nil {
+		return err
+	}
+
+	psact, err := client.PodSecurityAdmissionConfigurationTemplate.ByID(id)
+	if err != nil {
+		if IsNotFound(err) || IsForbidden(err) {
+			log.Printf("[INFO] PodSecurityAdmissionConfigurationTemplate with ID %s not found.", id)
```
Comment (bfbachmann) [inline rancher2/resource_rancher2_pod_security_admission_configuration_template.go:117]: Will we get the "forbidden" error if the PSACT is not found? This log suggests that's the case, but that seems strange.

File: rancher2/resource_rancher2_pod_security_admission_configuration_template_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  exemptions {
+    usernames = ["testuser"]
+    runtime_classes = ["testclass"]
+    namespaces = ["ingress-nginx","kube-system"]
+  }
+}
+`
+)
+
+func init() {}
```
Comment (bfbachmann) [inline rancher2/resource_rancher2_pod_security_admission_configuration_template_test.go]: Can we remove this since it's empty?

**Review by bfbachmann — COMMENTED** [review]
LGTM for the most part. Just small comments.

File: rancher2/resource_rancher2_pod_security_admission_configuration_template.go (hunk 2 above)
Comment (jiaqiluo) [reply rancher2/resource_rancher2_pod_security_admission_configuration_template.go:117]: TBH, I did not hit the forbidden error myself while testing, but it looks like a custom to check the notFound error along with the forbidden error in this repo.  To make it reasonable, my best guess is that there is a chance to hit the forbidden error if TF requests the resource while it is being removed on the server side.

File: rancher2/resource_rancher2_pod_security_admission_configuration_template_test.go (hunk 3 above)
Comment (jiaqiluo) [reply rancher2/resource_rancher2_pod_security_admission_configuration_template_test.go]: fixed.

File: rancher2/resource_rancher2_cluster_v2.go (hunk 1 above)
Comment (jiaqiluo) [reply rancher2/resource_rancher2_cluster_v2.go]: fixed.

**Review by jakefhyde — APPROVED** [review]

**Review by bfbachmann — APPROVED** [review]

**Pull request merged**